packages/ingest/granule.js

```diff
@@ -464,19 +470,60 @@ class Granule {
     return [null, null];
   }
 
+  /**
+   * rename s3 file with timestamp
+   *
+   * @param {string} bucket - bucket of the file
+   * @param {string} key - s3 key of the file
+   * @returns {Promise} promise that resolves when file is renamed
+   */
+  async renameS3FileWithTimestamp(bucket, key) {
+    const formatString = 'YYYYMMDDTHHmmssSSS';
+    let renamedKey = `${key}.v${moment.utc().format(formatString)}`;
+
+    // if the renamed file already exists, get a new name
+    // eslint-disable-next-line no-await-in-loop
+    while (await aws.s3ObjectExists({ Bucket: bucket, Key: renamedKey })) {
+      renamedKey = `${key}.v${moment.utc().format(formatString)}`;
+    }
+
+    log.debug(`renameS3FileWithTimestamp renaming ${bucket} ${key} to ${renamedKey}`);
+    return exports.moveGranuleFile(
+      { Bucket: bucket, Key: key }, { Bucket: bucket, Key: renamedKey }
+    );
+  }
+
+  /**
+   * get all renamed s3 files for a given bucket and key
+   *
+   * @param {string} bucket - bucket of the file
+   * @param {string} key - s3 key of the file
+   * @returns {Array<Object>} returns renamed files
+   */
+  async getRenamedS3File(bucket, key) {
+    const s3list = await aws.listS3ObjectsV2({ Bucket: bucket, Prefix: `${key}.v` });
+    return s3list.map((c) => ({ Bucket: bucket, Key: c.Key, fileSize: c.Size }));
+  }
+
   /**
    * Ingest individual files
    *
    * @private
    * @param {Object} file - file to download
    * @param {string} bucket - bucket to put file in
    * @param {string} duplicateHandling - how to handle duplicate files
-   * value can be `skip` to skip duplicates,
-   * or 'version' to create a new version of the file in s3
-   * @returns {Promise<Object>} returns promise that resolves to a file object
+   * value can be
+   * 'error' to throw an error,
+   * 'replace' to replace the duplicate,
+   * 'skip' to skip duplicate,
+   * 'version' to keep both files if they have different checksums
+   * @returns {Array<Object>} returns the staged file and the renamed existing duplicates if any
    */
   async ingestFile(file, bucket, duplicateHandling) {
     // Check if the file exists
+    let destinationKey = path.join(this.fileStagingDir, file.name);
+    if (destinationKey[0] === '/') destinationKey = destinationKey.substr(1);
```

Comment:
We should test this case in the unit tests

Reply:
Moved the check to constructor.  Added unit test.